File: ruby/space-age/space_age.rb
class SpaceAge

  EARTH_ORBITAL_PERIOD_SECONDS = 31557600.0

  ORBITAL_PERIODS = {
    :earth   => 1,
    :mercury => 0.2408467,
    :venus   => 0.61519726,
    :mars    => 1.8808158,
    :jupiter => 11.862615,
    :saturn  => 29.447498,
    :uranus  => 84.016846,
    :neptune => 164.79132,
  }

  ROUNDING = 2

  attr_reader :seconds

  def initialize(seconds)
    @seconds = seconds
  end

  ORBITAL_PERIODS.each do |planet, factor|

    define_method "on_#{planet}" do
      (calculate_age_on_earth / factor).round(ROUNDING)
    end

  end

  private

  def calculate_age_on_earth
    (@seconds / EARTH_ORBITAL_PERIOD_SECONDS)
  end

end

File: elixir/space-age/space_age.exs
defmodule SpaceAge do
  @type planet :: :mercury | :venus | :earth | :mars | :jupiter
                | :saturn | :neptune | :uranus

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet'.
  """
  @spec age_on(planet, pos_integer) :: float

  def age_on(:earth, seconds),   do: seconds / 31557600.0

  def age_on(:mercury, seconds), do: age_on(:earth, seconds) / 0.2408467
  def age_on(:venus, seconds),   do: age_on(:earth, seconds) / 0.61519726
  def age_on(:mars, seconds),    do: age_on(:earth, seconds) / 1.8808158
  def age_on(:jupiter, seconds), do: age_on(:earth, seconds) / 11.862615
  def age_on(:saturn, seconds),  do: age_on(:earth, seconds) / 29.447498
  def age_on(:uranus, seconds),  do: age_on(:earth, seconds) / 84.016846
  def age_on(:neptune, seconds), do: age_on(:earth, seconds) / 164.79132

end
